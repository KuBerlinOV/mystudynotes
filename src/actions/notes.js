import { v4 as uuidv4 } from 'uuid';

//ADD_NOTE
export const addNote = ({ topic = '', // in this case you should imagine destructuring as you are passing some object with the same data like ' const movie = {title, year, id} '
    description = '',           //when we are passing it to the action function as an argument and destructuring it. 
    note = '',                 //when it would be like ' const {title, year, id} = movie ' 
    reference = '',            //Furthermore, we provide here the default values for the object keys so
    tag = '#',
    status = 'in progress',     //the user could leave these options empty if he wanted to.
    createdAt = 0
} = {}) => {
    return {
        type: 'ADD_NOTE',
        note: {
            id: uuidv4(),
            topic,
            description,
            note,
            reference,
            tag,
            status,
            createdAt
        }
    }
}

//EDIT_NOTE

export const updateNote = (id, updates) => ({
    type: 'UPDATE_NOTE',
    id,
    updates
})

export const updateStatus = (id) => ({
    type: 'UPDATE_STATUS',
    id
})

//REMOVE_NOTE

export const removeNote = (id) => ({
    type: 'REMOVE_NOTE',
    id
})