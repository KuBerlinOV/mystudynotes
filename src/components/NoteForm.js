import React, { useState } from 'react';
import Modal from 'react-modal';
import moment from 'moment';
import { Link } from 'react-router-dom';

//moment 
const now = moment().format('MMMM Do YYYY, h:mm:ss a');
//props.noteBeforeUpdate ? moment(props.noteBeforeUpdate.createdAt) :

const NoteForm = (props) => {


    //local states
    const [state, setState] = useState({
        topic: props.noteBeforeUpdate ? props.noteBeforeUpdate.topic : '',
        description: props.noteBeforeUpdate ? props.noteBeforeUpdate.description : '',
        note: props.noteBeforeUpdate ? props.noteBeforeUpdate.note : '',
        reference: props.noteBeforeUpdate ? props.noteBeforeUpdate.reference : '',
        tag: props.noteBeforeUpdate ? props.noteBeforeUpdate.tag : '',
        createdAt: props.noteBeforeUpdate ? props.noteBeforeUpdate.createdAt : moment().format('MMMM Do YYYY, h:mm:ss a'),
        error: false
    });


    const handleChange = (e) => {
        setState({
            ...state,
            //this is the name of the element that we are targeting, depending on which input element
            //
            [e.target.name]: e.target.value
        })
    }

    const handleSubmit = (e) => {
        if (!state.topic || !state.description) {
            setState({ error: true })
        } else {
            setState({ error: false })
            //this is the props with the local state.
            //This local state then passed with the props
            // to notes component where the action 'ADD_NOTE' is dispatched with this local state
            props.handleSubmit(state)
        }
        e.preventDefault();
    }
    //explanation why this value={state.topic || ''} was needed https://stackoverflow.com/questions/44312437/react-router-v4-this-props-history-push-not-working
    return (
        <div>
            <form onSubmit={handleSubmit} action="">
                <h3>Create Note</h3>
                <h3>Topic:</h3>
                <input type="text" placeholder="topic" name="topic" autoFocus value={state.topic || ''} onChange={handleChange} />
                <h3>Description</h3>
                <input type="text" placeholder="description" name="description" id="" value={state.description || ''} onChange={handleChange} />
                <h3>Note</h3>
                <textarea placeholder="write your note here" name="note" id="" cols="30" rows="10" value={state.note || ''} onChange={handleChange} ></textarea>
                <h3>Reference</h3>
                <input type="text" placeholder="reference" name="reference" id="" value={state.reference || ''} onChange={handleChange} />
                <h3>Tag</h3>
                <input type="text" placeholder="tag" name="tag" value={state.tag || ''} onChange={handleChange} />

                <div>
                    {state.error && <p>Please provide topic and description</p>}
                    <button>Save</button>
                </div>
            </form>
        </div >
    )
}




export default NoteForm