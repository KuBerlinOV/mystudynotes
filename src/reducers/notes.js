//reducers = notes
const notesReducerDefaultState = []

const notesReducer = (state = notesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_NOTE':
            return [...state, action.note];
        case 'UPDATE_NOTE':
            return state.map(note => {
                if (note.id === action.id) {
                    return {
                        ...note,
                        ...action.updates
                    }
                } else {
                    return note
                }
            });
        case 'REMOVE_NOTE':
            return state.filter(({ id }) => id !== action.id);
        case 'UPDATE_STATUS':
            return state.map(note => {
                if (note.id === action.id && note.status === 'in progress') {
                    return {
                        ...note,
                        status: 'mastered'
                    }
                } else if (note.id === action.id && note.status === 'mastered') {
                    return {
                        ...note,
                        status: 'in progress'
                    }
                } else {
                    return {
                        ...note
                    }
                }
            })
        default:
            return state;
    }

}

export default notesReducer;
