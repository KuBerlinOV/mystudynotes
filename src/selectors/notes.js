


const selectNotes = (notes, { text, sortBy, startDate, endDate }) => {
    return notes.filter(note => {
        const descriptionToMatch = note.description.toLowerCase()
        const topicToMatch = note.topic.toLowerCase();
        const textToMatch = text.toLowerCase();

        const textMatch = descriptionToMatch.includes(textToMatch) || topicToMatch.includes(textToMatch) ? true : false;
        const startDateMatch = typeof startDate !== 'number' || note.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || note.createdAt <= endDate;

        return textMatch && startDateMatch && endDateMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'status') {
            return a.status === 'mastered' ? 1 : -1
        }
    })
}

export default selectNotes;