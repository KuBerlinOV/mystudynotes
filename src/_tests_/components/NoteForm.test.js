import React from 'react';
import { shallow } from 'enzyme';
import NoteForm from '../../components/NoteForm';
import notes from '../fixtures/notes';
import { render, fireEvent, cleanup } from '@testing-library/react';

describe('NoteForm', () => {
    it('should render NoteForm correctly', () => {
        const wrapper = shallow(<NoteForm />)
        expect(wrapper).toMatchSnapshot();
    })
    it('should render NoteForm with a note data correctly', () => {
        const wrapper = shallow(<NoteForm noteBeforeUpdate={notes[0]} />)
        expect(wrapper).toMatchSnapshot();
    })
    it('should render error for invalid form submission', () => {
        const wrapper = shallow(<NoteForm />)
        expect(wrapper).toMatchSnapshot()
        wrapper.find('form').simulate('submit', {
            preventDefault: () => { }
        });
        expect(wrapper.findByProps('error')).toBeTruthy()// <--this simulates that the 
        expect(wrapper).toMatchSnapshot()
    })
})
