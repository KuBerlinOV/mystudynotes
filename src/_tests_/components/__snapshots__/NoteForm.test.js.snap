// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteForm should render NoteForm correctly 1`] = `
<div>
  <div>
    <form
      action=""
      data-testid="form"
    >
      <h3>
        Create Note
      </h3>
      <h3>
        Topic:
      </h3>
      <input
        name="topic"
        placeholder="topic"
        type="text"
        value=""
      />
      <h3>
        Description
      </h3>
      <input
        id=""
        name="description"
        placeholder="description"
        type="text"
        value=""
      />
      <h3>
        Note
      </h3>
      <textarea
        cols="30"
        id=""
        name="note"
        placeholder="write your note here"
        rows="10"
      />
      <h3>
        Reference
      </h3>
      <input
        id=""
        name="reference"
        placeholder="reference"
        type="text"
        value=""
      />
      <h3>
        Tag
      </h3>
      <input
        name="tag"
        placeholder="tag"
        type="text"
        value=""
      />
      <div>
        
        <button>
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`NoteForm should render NoteForm with a note data correctly 1`] = `
<div>
  <div>
    <form
      action=""
      data-testid="form"
    >
      <h3>
        Create Note
      </h3>
      <h3>
        Topic:
      </h3>
      <input
        name="topic"
        placeholder="topic"
        type="text"
        value="redux"
      />
      <h3>
        Description
      </h3>
      <input
        id=""
        name="description"
        placeholder="description"
        type="text"
        value="redux note"
      />
      <h3>
        Note
      </h3>
      <textarea
        cols="30"
        id=""
        name="note"
        placeholder="write your note here"
        rows="10"
      >
        here is a note
      </textarea>
      <h3>
        Reference
      </h3>
      <input
        id=""
        name="reference"
        placeholder="reference"
        type="text"
        value="some reference"
      />
      <h3>
        Tag
      </h3>
      <input
        name="tag"
        placeholder="tag"
        type="text"
        value="some tag"
      />
      <div>
        
        <button>
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`NoteForm should render error for invalid form submission 1`] = `
<div>
  <div>
    <form
      action=""
      data-testid="form"
    >
      <h3>
        Create Note
      </h3>
      <h3>
        Topic:
      </h3>
      <input
        name="topic"
        placeholder="topic"
        type="text"
        value=""
      />
      <h3>
        Description
      </h3>
      <input
        id=""
        name="description"
        placeholder="description"
        type="text"
        value=""
      />
      <h3>
        Note
      </h3>
      <textarea
        cols="30"
        id=""
        name="note"
        placeholder="write your note here"
        rows="10"
      />
      <h3>
        Reference
      </h3>
      <input
        id=""
        name="reference"
        placeholder="reference"
        type="text"
        value=""
      />
      <h3>
        Tag
      </h3>
      <input
        name="tag"
        placeholder="tag"
        type="text"
        value=""
      />
      <div>
        
        <button>
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`NoteForm should render error for invalid form submission 2`] = `
<div>
  <div>
    <form
      action=""
      data-testid="form"
    >
      <h3>
        Create Note
      </h3>
      <h3>
        Topic:
      </h3>
      <input
        name="topic"
        placeholder="topic"
        type="text"
        value=""
      />
      <h3>
        Description
      </h3>
      <input
        id=""
        name="description"
        placeholder="description"
        type="text"
        value=""
      />
      <h3>
        Note
      </h3>
      <textarea
        cols="30"
        id=""
        name="note"
        placeholder="write your note here"
        rows="10"
      />
      <h3>
        Reference
      </h3>
      <input
        id=""
        name="reference"
        placeholder="reference"
        type="text"
        value=""
      />
      <h3>
        Tag
      </h3>
      <input
        name="tag"
        placeholder="tag"
        type="text"
        value=""
      />
      <div>
        <p>
          Please provide topic and description
        </p>
        <button>
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`NoteForm should update description input value on change 1`] = `
<div>
  <div>
    <form
      action=""
      data-testid="form"
    >
      <h3>
        Create Note
      </h3>
      <h3>
        Topic:
      </h3>
      <input
        name="topic"
        placeholder="topic"
        type="text"
        value=""
      />
      <h3>
        Description
      </h3>
      <input
        id=""
        name="description"
        placeholder="description"
        type="text"
        value="new description"
      />
      <h3>
        Note
      </h3>
      <textarea
        cols="30"
        id=""
        name="note"
        placeholder="write your note here"
        rows="10"
      />
      <h3>
        Reference
      </h3>
      <input
        id=""
        name="reference"
        placeholder="reference"
        type="text"
        value=""
      />
      <h3>
        Tag
      </h3>
      <input
        name="tag"
        placeholder="tag"
        type="text"
        value=""
      />
      <div>
        
        <button>
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`NoteForm should update note input value on change 1`] = `
<div>
  <div>
    <form
      action=""
      data-testid="form"
    >
      <h3>
        Create Note
      </h3>
      <h3>
        Topic:
      </h3>
      <input
        name="topic"
        placeholder="topic"
        type="text"
        value=""
      />
      <h3>
        Description
      </h3>
      <input
        id=""
        name="description"
        placeholder="description"
        type="text"
        value=""
      />
      <h3>
        Note
      </h3>
      <textarea
        cols="30"
        id=""
        name="note"
        placeholder="write your note here"
        rows="10"
      >
        new note
      </textarea>
      <h3>
        Reference
      </h3>
      <input
        id=""
        name="reference"
        placeholder="reference"
        type="text"
        value=""
      />
      <h3>
        Tag
      </h3>
      <input
        name="tag"
        placeholder="tag"
        type="text"
        value=""
      />
      <div>
        
        <button>
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`NoteForm should update topic input value on change 1`] = `
<div>
  <div>
    <form
      action=""
      data-testid="form"
    >
      <h3>
        Create Note
      </h3>
      <h3>
        Topic:
      </h3>
      <input
        name="topic"
        placeholder="topic"
        type="text"
        value="new topic"
      />
      <h3>
        Description
      </h3>
      <input
        id=""
        name="description"
        placeholder="description"
        type="text"
        value=""
      />
      <h3>
        Note
      </h3>
      <textarea
        cols="30"
        id=""
        name="note"
        placeholder="write your note here"
        rows="10"
      />
      <h3>
        Reference
      </h3>
      <input
        id=""
        name="reference"
        placeholder="reference"
        type="text"
        value=""
      />
      <h3>
        Tag
      </h3>
      <input
        name="tag"
        placeholder="tag"
        type="text"
        value=""
      />
      <div>
        
        <button>
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;
